cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -10,70 +10,302 @@
+        String vId = context.namingStrategy().getId(CgmesNames.VOLTAGE_LEVEL, voltageLevelId);
+        voltageLevelAdjacency.computeIfAbsent(vId, k -> new ArrayList<>());
+    }
+
+    private void addSubstation(Map<String, List<String>> substationAdjacency, PropertyBag sub) {
+        String substationlId = sub.getId(CgmesNames.SUBSTATION);
+        String subId = context.namingStrategy().getId(CgmesNames.SUBSTATION, substationlId);
+        substationAdjacency.computeIfAbsent(subId, k -> new ArrayList<>());
+    }
+
+    // Two different voltageLevels are adjacent if they are connected by a switch
+    // If the corresponding substations are different they are also adjacent
+    private void addSwitch(Map<String, List<String>> voltageLevelAdjacency,
+        Map<String, List<String>> substationAdjacency, PropertyBag sw) {
+
+        CgmesTerminal t1 = context.cgmes().terminal(sw.getId(CgmesNames.TERMINAL + 1));
+        String node1 = context.nodeBreaker() ? t1.connectivityNode() : t1.topologicalNode();
+        String voltageLevelId1 = getVoltageLevelFromNode(node1, t1);
+
+        CgmesTerminal t2 = context.cgmes().terminal(sw.getId(CgmesNames.TERMINAL + 2));
+        String node2 = context.nodeBreaker() ? t2.connectivityNode() : t2.topologicalNode();
+        String voltageLevelId2 = getVoltageLevelFromNode(node2, t2);
+
+        // Null could be received as voltageLevel at the boundary
+        if (voltageLevelId1 == null || voltageLevelId2 == null || voltageLevelId1.equals(voltageLevelId2)) {
+            return;
+        }
+        addAdjacency(voltageLevelAdjacency, voltageLevelId1, voltageLevelId2);
+
+        String substationId1 = context.cgmes().substation(t1, context.nodeBreaker());
+        String substationId2 = context.cgmes().substation(t2, context.nodeBreaker());
+
+        // Null could be received as substation at the boundary
+        if (substationId1 == null || substationId2 == null || substationId1.equals(substationId2)) {
+            return;
+        }
+        addAdjacency(substationAdjacency, substationId1, substationId2);
+    }
+
+    private String getVoltageLevelFromNode(String node, CgmesTerminal t) {
+        String voltageLevelId = null;
+        if (node != null && !context.boundary().containsNode(node)) {
+            voltageLevelId = context.cgmes().voltageLevel(t, context.nodeBreaker());
+        }
+        return voltageLevelId;
+    }
+
+    // Two different substations are adjacent if they are connected by a transformer
+    private void addEnds(Map<String, List<String>> substationAdjacency, PropertyBags tends) {
+        List<String> substationsIds = substationsIds(tends);
+        if (substationsIds.size() <= 1) {
+            return;
+        }
+        String sub0 = substationsIds.get(0);
+        for (int i = 1; i < substationsIds.size(); i++) {
+            String subi = substationsIds.get(i);
+
+            if (sub0.contentEquals(subi)) {
```

Comment:
If sub(0)'s content is equal to sub(i)'s content, that means the two substations are the same? Why isn't sufficient enough to put the substations IDs in a set and call addAdjacency for all couple?

When `substationsIds` contains three different values (3WT), I'm not sure your proposal works well: it creates the following association (0,1), (1, 0), (0, 2) and (2, 0) but not (1, 2) and (2, 1). Is it correct?

Reply:
The code is ok. We only need to connect the three substations (0, 1, and 2), so associating the substation 0 with the substations 1 and 2 is enough.